pkg/client/constructor.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+	"k8s.io/client-go/rest"
+)
+
+type ClientConstructor[R any] struct {
+	NewForConfigAndClient func(*rest.Config, *http.Client) (R, error)
+}
+
+type ClientCache[R any] interface {
+	ClusterOrDie(name logicalcluster.Name) R
+	Cluster(name logicalcluster.Name) (R, error)
+}
+
+func NewClientCache[R any](cfg *rest.Config, client *http.Client, constructor *ClientConstructor[R]) ClientCache[R] {
+	return &clientCache[R]{
+		cfg:         cfg,
+		client:      client,
+		constructor: constructor,
+
+		RWMutex:          &sync.RWMutex{},
+		clientsByCluster: map[logicalcluster.Name]R{},
+	}
+}
+
+type clientCache[R any] struct {
+	cfg         *rest.Config
+	client      *http.Client
+	constructor *ClientConstructor[R]
+
+	*sync.RWMutex
```

Comment:
Does this need to be a pointer?

Reply:
IIRC it gets passed around, and can't be by value, right?

Reply:
If it gets passed around, then yes, pointer it is